### PR TITLE
Adding EnumerateChunks which allow efficient scanning of a StringBuilder

### PR DESCRIPTION
**Pull request by vancem**
This change is the start of addressing the issue  CoreFX issue 

https://github.com/dotnet/corefx/issues/21248

which wants to have an efficient way of reading the characters of a StringBuilder.      However it has morphed alot since then (from a callback to an iterator model).   The API review part of this is the following issue. 

https://github.com/dotnet/corefx/issues/29770

This PR is an implementation of that spec.  In particular we implement the following new method on StringBuilder. 
```
class StringBuilder { 
      ChunkEnumerator EnumerateChunks();
}
```
ChunkEnumerator is a struct that follows the IEnumerable<ReadOnlyMemory<char>> pattern and thus works with the C# foreach statement.   Here is an example for efficiently implementing 'IndexOf' using EnumerateChunks
```
        static int IndexOf(StringBuilder sb, char c)
        {
            int pos = 0;
            foreach (ReadOnlyMemory<char> chunk in sb.EnumerateChunks())
            {
                var span = chunk.Span;  // We create a local variable because it is more efficient
                for (int i = 0; i < span .Length; i++)
                    if (span [i] == c)
                        return pos + i;
                pos += span.Length;
            }
            return -1;
        }
```
The PR contains the actual implementation.    Note that ChunkEnumerable does not actually implement IEnumerable<ReadOnlyMemory<char>>.  

We considered having it return ReadOnlySpan, but decided that it was too fragile in that Span can't be used across 'await' calls and can't be passed to methods that MAY be async.   ReadOnlyMemory does not have this problem (but is a bit less efficient), but does require you to convert it to a Span before accessing the characters (and notice you need to cache the span in a local variable to keep span creation out of the inner loop).    

This is not ideal, but is this issue will resurface other places (whenever we want to return chunks), and so it is likely we will invest in JIT optimization to optimize away any inefficiency of using ReadOnlyMemory and converting immediately to Span (which is what would happen in the synchronous case).  
 
At this point it is ready for API review.    This amounts to signing off on using ReadOnlyMemory instead of ReadOnlySpan, and the names (is 'Chunks' the right suffix?) and the fact that ChunkEnumerable does not actually implement IEnumerable<ReadOnlyMemory<char>> (It could, it is just more code (which we can add later)).  

I have tests, however there seems to be a chicken and egg problem in that you need to have an implementation, before CoreFX can expose it (and have tests for it).    I have tested the code (all code paths are covered), but I will need to check in tests as a separate PR.

### CODE REVIEW

**Comment by vancem** [comment]
@jaredpar  @VSadov - above is a proposal for a way of getting chunks out of an StringBuilder.   It is very simple, it is an Enumerator that returns ReadOnlySpan<char>.    

The stack-only ness of Span DOES NOT PREVENT building an Enumerator that returns it, but at least currently it would cause some grief if you try to use it in an async method.   Consider this example where i put some awaits in various points.  
```
        static int IndexOf(StringBuilder sb, char c)
        {
            int pos = 0;
            foreach (ReadOnlySpan<char> span in sb.EnumerateChunks())
            {
                await Task.Delay(10);  // Point 1
                for (int i = 0; i < span.Length; i++)
                    if (span[i] == c)
                        await Task.Delay(10);  // Point 2
                pos += span.Length;
                await Task.Delay(10);  // Point 3
            }
            return -1;
        }
```
Today only Point 3 would be allowed as the span is 'dead' at that point and thus it would not need to be captured.  

However the compiler could be smart, and frankly after every 'await' it could insert a 'span=enumerator.Current' right after every 'await' and NOT have to capture the span into the closure.

The question to you is :   What do you think about this as a feature?    If you like it, how hard is it?   

While this feature would be handy for this particular API, it really is a general problem of enumerating a collection of spans (in an async environment).    We can fix it just by avoiding spans (using memory instead), but that is both less efficient, and more semantically complex.   

What do you think?  (Today people can work-around this issue by expanding the 'foreach' by hand, but that is reasonably ugly....)

cc: @davidfowl, @stephentoub, @jkotas

**Comment by vancem** [comment]
@jaredpar 

I think part of the makes this idea tractable is that there is a strong expectation that Enumerator.Current is side-effect free.     Thus you could simply state that it is undefined what happens when this is not true (that is the compiler is allowed to insert as many calls to Enumerator.Current as it wants).    You could narrow this if you like to say that this freedom only applies to enumerators that return ref-like structs and use await.   

Alternatively, you could say that semantics is that yes, logically there is EXACTLY a call to Enumerator.Current at variable initialization, and then after every await statement, but the compiler is free to optimize these away if it can.   To first order, it simply does not (the JIT may or may not do it).    This is a bit sub-optimal, but I could easily live with it.  

I want to note, however that the CURRENT behavior is pretty unhelpful (if you try to use enumerations of Span in an async method, you will get errors that you can't do that (even though there is an obvious way to make it work).    This is rather unfortunate. 

If it were me I would define the semantics such that compiler is free to call Enumerator.Current as many times as it wants (that is it is the behavior can vary if Enumerator.Current has side effects).

**Comment by jaredpar** [comment]
> I think part of the makes this idea tractable is that there is a strong expectation that Enumerator.Current is side-effect free. 

I do not share that expectation. Customers do bad things with their code all the time. This wouldn't be the first, second, third, etc ... time that a method we strongly expected to be side effect free was indeed not side effect free. 

> Thus you could simply state that it is undefined what happens when this is not true (that is the compiler is allowed to insert as many calls to Enumerator.Current as it wants).

Undefined behavior is for C++. 😄

**Comment by vancem** [comment]
If you really want to have well defined semantics in the face of Enumerator.Current having side effects, you can do that.  Simply state it happens when the variable is created, and after every await.   You are done.  Perfectly well defined semantics, AND semantics that people WANT (because it turns error behavior into very reasonable non-error behavior).  

The only reason the compile does this (and not the user), is because the user CAN'T because the COMPILER hid the variable needed to do it.

**Comment by VSadov** [comment]
The code does not look too bad, but it would not currently compile. 

There is a bigger problem. Any kind of Span locals are not allowed in async methods right now. There is a proposal to relax this for `{  }` blocks that do not contain awaits. 

The reason why we do not want to key of the capturing is because capturing is an implementation detail. We keep improving/changing the capturing story with every version. As the strategy changes we may capture less (or sometimes more, but hopefully less). Also the strategy is different between Release and Debug. In Debug we capture more aggressively, otherwise locals may vanish after `await`.

For once I cannot guaranty that Mono C# compiler captures exactly the same set of locals. It most certainly does not.

Basically we do not want to rely on a fluid thing full of optional implementation details and optimizations as a reason to accept or reject programs.

**Comment by VSadov** [comment]
That would still run into the issue with Span locals not being allowed in async methods, but assuming https://github.com/dotnet/csharplang/issues/1331 , it could work.

**Comment by vancem** [comment]
The following code should work today but it needs a helper method. 
Basically it passes the span into the helper and the helper does all the work, and returns if it needs to await.    This could get arbitrarily ugly to partition your code in this way, but I see that realistic cases are likely to be fine (it is not a hardship.   Indeed with https://github.com/dotnet/csharplang/issues/1331 you probably would not need the helper method most of the time and the code would still be clean.  
```
        static async Task<int> IndexOf1(StringBuilder sb, char c)
        {
            var spanEnum = sb.EnumerateChunks().GetEnumerator();
            while (spanEnum.MoveNext())
            {
                await Task.Delay(10); // Point 1

                int i = 0;
                while (Helper(spanEnum.Current, char c))
                    await Task.Delay(10); // Point 2

                await Task.Delay(10); //Point 3
            }
            return -1;
        }

        static bool Helper(ReadOnlySpan<char> span, ref int i, char c)
        {
            while (i < span.Length)
            {
                if (span[i] == c)
                    return true;
                i++;
            }
            return false;
        }
```
So I think the upshot is that you CAN make things work with an Enumerable of a ReadOnlySpan<Char> in an async method, but https://github.com/dotnet/csharplang/issues/1331 makes it better in most cases.   It does mean that you can't use foreach over it in any async method.    This may be acceptable.

**Comment by VSadov** [comment]
BTW: You can make `Helper` a local nested method and get very close (syntactically, and operationally, if JIT inlines it) to what dotnet/csharplang#1331 promises.

**Comment by vancem** [comment]
Ah, I do see that variable capture inside the nested function is pretty nice here.  The main pain is that you have to return to do any await, which is a pain, I am not at all convinced that it will happen enough to matter at all.

At this point I think I am becoming comfortable with the API returning an Enumeratable over ReadOnlySpan<char> (we don't need ReadOnlyMemory).   It works great for synchronous code, and it is not that bad for async code.

**Comment by davidfowl** [comment]
@vancem what if I want to call an asynchronous TextWriter method?

**Comment by danmosemsft** [comment]
@vancem can you clarify what the "chunks" are in this context? Do they map to the chunks in the implementation? That would couple API to implementation (which could change again, indeed StringBuilder was not originally chunk based) and also produce Spans of possibly wildly (and unexpectedly) varying sizes, depending on the history of the StringBuilder. Clearly the span returned cannot straddle a chunk boundary without an allocation somewhere. But would it make sense to cap the size of the span, if only to logically separate the implementation detail?

**Comment by VSadov** [comment]
I think the pattern is more general - a similar approach would work for anything backed by segmented buffers. Of which a "single buffer" is just a subset.

I think varying in sizes of the buffer is ok. 
It should probably never return an empty buffer (just because that is silly, but not dangerous really)

**Comment by vancem** [comment]
My response is basically the same as @VSadov.    All we want is an API that allows things to give back the characters efficiently in a way that is mostly independent of the implementation.    This proposal covers a broad swath (as long as the characters are stored as bobs of characters you can implement this interface efficiently).   Sure if you encode the stringbuilder as UTF8 chars you have a problem, but it is only a perf problem and it is no worse then the perf problem they have today (since today we force them to form a string.

**Comment by MohammadHamdyGhanem** [comment]
@vancem 
I hope you look at this and see if you can make use of some ideas:
Add a MutableString Class that contains the missing StringBuilder functinality  #29510 
Note that even if you improve the Indexer, using a for loop with StringBulider is still bad, becuase it iterates all the chunks from the beginning for each index. So, you must also make StringBuilder implement IEnumerator to allow us to loop using foreach, which uses MoveXX methods to go from current Index to the next, moving from one chunk to the next, without the need to start over from with every index.

**Comment by MohammadHamdyGhanem** [comment]
@vancem 
Why do you need a span or a memory to do enumeration?.. Just pass through each chunk and throw each char, using two variables to hold the current chunk and current charIndex. But this will need that each chunk has another ref to its next chunk not just the previous. This is a smaal price as chunkNo = SB.Length/8000.

**Comment by stephentoub** [comment]
@vancem, I think this should yield `ReadOnlyMemory<char>` rather than `ReadOnlySpan<char>`.  It's not just `await`'ing between iterations that's an issue, but rather being able to hand off the value to an asynchronous method as well, e.g.
```C#
foreach (ReadOnlyMemory<char> buffer in sb.EnumerateChunks())
{
    await textWriter.WriteAsync(buffer);
}
```
I can't do that if `buffer` is a span.

**Comment by vancem** [comment]
I have changed it from ReadOnlySpan back to ReadOnlyMemory.    It is fragile being a Span, and the cost of going to Memory is not high.

**Comment by vancem** [comment]
After a bunch of discussion above, I have decided to use ReadOnlyMemory<char> rather than ReadOnlySpan<char> because it allows the chunks to be passed around to APIs that are ASYNC (which can't use Span) (this is Stephen's point above).

The disadvantage of this choice is that 
1. Generally you want to access the characters, which means calling chunk.Span. Thus we made things slightly harder in a common case (but we allow more schenarios.
2. This is less efficient since you have to make a Memory object just to fetch a span from it and throw it away.   JIT Optimization could fix this however.

File: src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs [hunk 1]
```diff
@@ -557,6 +557,132 @@ public int Length
...
+        {
+            /// <summary>
+            /// Implements the IEnumerable pattern.  
+            /// </summary>
+            public ChunkEnumerator GetEnumerator() => new ChunkEnumerator(this._stringBuilder);
+
+            #region private
+            internal ChunkEnumerable(StringBuilder stringBuilder)
+            {
+                Debug.Assert(stringBuilder != null);    // Because it is only called with at 'this' pointer.
```
Comment (danmosemsft) [inline src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs]: Comment grammar?

File: src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs [hunk 2]
```diff
@@ -557,6 +557,132 @@ public int Length
...
+        /// as chunks (ReadOnlyMemory) of characters.  An example use is:
+        /// 
+        ///      foreach (ReadOnlyMemory<char> chunk in sb.EnumerateChunks())
+        ///      {  /* operate on chars using 'chunk' */   }
+        ///      
+        /// </summary>
+        public ChunkEnumerable EnumerateChunks() => new ChunkEnumerable(this);
+
+        /// <summary>
+        /// ChunkEnumerable supports the IEnumerable pattern so foreach works (see EnumerateChunks)
```
Comment (danmosemsft) [inline src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs]: Worth explaining here why you didn't make String builder implement IEnumerable? (Now you moved off Span, you could?)

File: src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs [hunk 3]
```diff
@@ -557,6 +557,132 @@ public int Length
...
+                if (_manyChunks != null)
+                    return _manyChunks.MoveNext(ref _currentChunk);
+
+                StringBuilder next = _firstChunk;
+                while (next.m_ChunkPrevious != _currentChunk)
+                    next = next.m_ChunkPrevious;
+                _currentChunk = next;
+                return true;
+            }
+            /// <summary>
```
Comment (danmosemsft) [inline src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs]: Nit newline above

File: src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs [hunk 4]
```diff
@@ -557,6 +557,132 @@ public int Length
...
+            internal ChunkEnumerator(StringBuilder stringBuilder)
+            {
+                Debug.Assert(stringBuilder != null);
+                _firstChunk = stringBuilder;
+                _currentChunk = null;   // MoveNext will find the last chunk if we do this.  
+                _manyChunks = null;
+
+                // There is a performance-vs-allocation tradeoff.   Because the chunks
+                // are a linked list with each chunk pointing to its PREDECESSOR, walking
+                // the list FORWARD is not efficient.   If there are few chunks (< 8) we
```
Comment (danmosemsft) [inline src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs]: I guess the number 8 is just a judgement call here? I don't have good intuition for the number of chunks in the real world.

File: src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs [hunk 5]
```diff
@@ -557,6 +557,132 @@ public int Length
...
+                // simply scan from the start each time, and tolerate the N*N behavior. 
+                // However above this size, we allocate an array to hold pointers to all
+                // the chunks and we can be efficient for large N.    
+                int chunkCount = ChunkCount(stringBuilder);
+                if (8 < chunkCount)
```
Comment (danmosemsft) [inline src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs]: Nit, this is sort of a Yoda conditional

File: src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs (hunk 2 above)
Comment (vancem) [reply src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs]: There is not a strong barrier to implementing IEnumerable.    The main reason for not doing so is that it does incur some cost in complexity (more methods (that mostly never get called) and testing, and does not seem to add important value.  Its main potential value is that you could pass it to APIs that take a IEnumerable, but that definitely would decrease the performance (you would end up boxing the struct), and chunks are really not a semantic meaningful quantity (it is note the natural thing you would do Linq over).   I don't we should be encouraging using Linq over chunks.  

So while it may have some value (which is why this question raised in https://github.com/dotnet/corefx/issues/29770), given that we could add it later compatibly, it is lower risk not to include it.

I hesitate to modify this comment however, as it is public facing documentation (and probably should avoid design discussions (which are better in PRs and Issues).

File: src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs (hunk 4 above)
Comment (vancem) [reply src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs]: Yes, 8 is a judgement call.   Bascially you want a number before N*N gets too big.     Generally if you do 'nothing' StringBuilders start with 16 bytes and double each time, so this handles strings up to about 2K-4K if you don't ever specify a starting length.    Thus 8 keeps things non-allocating for the common case of strings < 256 bytes (e.g. paths).     On the other hand, at 2-4K of data, the cost of copying the data is probably now large with respect to the cost of allocating the array, so the value of making the nubmer bigger is probably not a big deal (especially because people using StringBuilder on large strings should be setting a good starting length).  

We can tune it as necessary.

File: src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs (hunk 5 above)
Comment (vancem) [reply src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs]: I don't follow the change you want.

Comment (davidfowl) [reply src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs]: `if (chunkCount > 8)` https://en.wikipedia.org/wiki/Yoda_conditions

Comment (vancem) [reply src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs]: I see.   Quite a while back, I adopted a convention of always making relational operators go from smallest to largest.   (basically I never use > or >=).    I find it slightly clearer and less error prone, as it eliminates one source of arbitrary variation, and allows visual position to help in reading the code (this is especially nice if you are testing for a range).   

For what it is worth...

File: src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs [hunk 6]
```diff
@@ -557,6 +557,145 @@ public int Length
...
+            /// </summary>
+            public ChunkEnumerator GetEnumerator() => new ChunkEnumerator(this._stringBuilder);
+
+            #region private
+            internal ChunkEnumerable(StringBuilder stringBuilder)
+            {
+                Debug.Assert(stringBuilder != null);    // Because it is only called with a 'this' pointer.    
+                _stringBuilder = stringBuilder;
+            }
+            private StringBuilder _stringBuilder;   // We insure this is never null.
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs]: Nit: insure => ensure

Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs]: Nit: readonly

File: src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs [hunk 7]
```diff
@@ -557,6 +557,145 @@ public int Length
...
+        ///    }
+        /// </summary>
+        public ChunkEnumerable EnumerateChunks() => new ChunkEnumerable(this);
+
+        /// <summary>
+        /// ChunkEnumerable supports the IEnumerable pattern so foreach works (see EnumerateChunks)  
+        /// It needs to be public (so the compiler can use it when building a foreach statement) 
+        /// but users typically don't use it explicitly (which is why it is a nested type).  
+        /// </summary>
+        public struct ChunkEnumerable
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs]: Nit: readonly?

File: src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs [hunk 8]
```diff
@@ -557,6 +557,145 @@ public int Length
...
+                    _chunks = new StringBuilder[chunkCount];
+                    while (0 <= --chunkCount)
+                    {
+                        Debug.Assert(stringBuilder != null);
+                        _chunks[chunkCount] = stringBuilder;
+                        stringBuilder = stringBuilder.m_ChunkPrevious;
+                    }
+                    _chunkPos = -1;
+                }
+                StringBuilder[] _chunks;    // These are in normal order (first chunk first)
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs]: Nit: private readonly

File: src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs (hunk 5 above)
Comment (danmosemsft) [reply src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs]: That makes sense. I may do it.

**Comment by vancem** [comment]
OK, I have made the updates based on the API review https://github.com/dotnet/corefx/issues/29770, so now we just have ChunkEnumerator, but from a users point of view (that he can use foreach) nothing has changed.  

I have tests to add, but they go in the CoreFX repo along with changes to System.Runtime, and have to be done after this merge is complete and we have a good build that CoreFX can refer to.   

So fair warning: I am going to merge this later today (so we tonight's build has it), so if you have any reservations, speak now (in the next hour or two).
